// @jbcom/strata/dist/api/effects.js
/**
 * @module Effects
 * @category Effects & Atmosphere
 *
 * Effects & Atmosphere - Particles, Weather, Lighting, and Decals
 *
 * Visual effects that bring your world to life - from rain and snow
 * to explosions and god rays.
 *
 * @example
 * ```tsx
 * import { ParticleEmitter, Rain, GodRays } from '@jbcom/strata/api/effects';
 *
 * function WeatherScene() {
 *   return (
 *     <>
 *       <Rain intensity={0.5} />
 *       <GodRays lightPosition={[10, 20, 0]} />
 *       <ParticleEmitter preset="fire" position={[0, 0, 0]} />
 *     </>
 *   );
 * }
 * ```
 */
// Particle Systems - React components
// Weather Effects - React components
// Decals & Billboards - React components
// LOD System - React components
// God Rays & Volumetric Lighting - React components
export { AnimatedBillboard, Billboard, Decal, DecalPool, GodRays, Impostor, Lightning, LightShafts, LODGroup, LODMesh, LODVegetation, ParticleBurst, ParticleEmitter, Rain, Snow, VolumetricPointLight, VolumetricSpotlight, WeatherEffects, } from '../components';
// Particle Systems - Core utilities
// Weather Effects - Core utilities
// Decals & Billboards - Core utilities
// LOD System - Core utilities
// God Rays & Volumetric Lighting - Core utilities
export { applySpriteSheetFrame, batchLODObjects, blendGodRayColors, calculateGodRayIntensityFromAngle, calculateImpostorAngle, calculateLODLevel, calculateRadialBlur, calculateScatteringIntensity, calculateScreenSpaceSize, calculateSunOcclusion, calculateTemperature, calculateVegetationDensity, createBillboardMatrix, createBloodSplatterTexture, createBulletHoleTexture, createDecalTexture, createDitherPattern, createFootprintTexture, createGodRaysMaterial, createImpostorGeometry, createImpostorTexture, createLODLevels, createParticleEmitter, createPointLightSphereGeometry, createScorchMarkTexture, createSpotlightConeGeometry, createSpriteSheetAnimation, createSpriteSheetMaterial, createVegetationLODLevels, createVolumetricPointLightMaterial, createVolumetricSpotlightMaterial, createWaterPuddleTexture, createWeatherSystem, createWindSimulation, DecalProjector, generateLODGeometries, getLightScreenPosition, getPrecipitationType, getSpriteSheetUVs, interpolateLODMaterials, LODManager, ParticleEmitter as CoreParticleEmitter, shouldUseLOD, simplifyGeometry, sortBillboardsByDepth, updateBillboardRotation, updateGodRaysLightPosition, updateImpostorUV, updateSpriteSheetAnimation, WeatherSystem as CoreWeatherSystem, WindSimulation, } from '../core';
