// @jbcom/strata/dist/api/experience.js
/**
 * @module Experience
 * @category Player Experience
 *
 * Player Experience - Cameras, Input, Audio, and UI
 *
 * Everything that connects the player to your game - how they see it,
 * how they control it, what they hear, and what information they receive.
 *
 * @example
 * ```tsx
 * import { FollowCamera, AudioProvider, HealthBar } from '@jbcom/strata/api/experience';
 *
 * function Game() {
 *   return (
 *     <AudioProvider>
 *       <FollowCamera target={playerRef} distance={5} />
 *       <HealthBar current={75} max={100} />
 *     </AudioProvider>
 *   );
 * }
 * ```
 */
// Camera Systems - React components
// Input Handling - React components
// Audio System - React components
// Game UI - React components
export { AmbientAudio, AudioEmitter, AudioEnvironment, AudioListener, AudioProvider, AudioZone, CameraShake, CinematicCamera, Crosshair, DamageNumber, DialogBox, FollowCamera, FootstepAudio, FPSCamera, GroundSwitch, HealthBar, Inventory, Joystick3D, Minimap, Nameplate, Notification, OrbitCamera, PositionalAudio, PressurePlate, ProgressBar3D, Tooltip, TriggerComposer, useAudioContext, useAudioManager, useCameraTransition, WallButton, WeatherAudio, } from '../components';
// Camera Systems - Core utilities
// Input Handling - Core utilities
// Audio System - Core utilities
// Game UI - Core utilities
export { angleToAxis, axisToAngle, axisToMagnitude, CameraShake as CoreCameraShake, calculateFade, calculateHeadBob, calculateLookAhead, calculateScreenShakeIntensity, clampAxis, clampProgress, createDefaultCrosshair, createDefaultDamageNumber, createDefaultDialog, createDefaultInventory, createDefaultMinimap, createDefaultNameplate, createDefaultNotification, createDefaultProgressBar, createDefaultTooltip, createInputManager, createSoundManager, createSpatialAudio, DEFAULT_SPATIAL_CONFIG, ENVIRONMENT_PRESETS, easeInCubic, easeInOutCubic, easeOutCubic, easeOutElastic, evaluateCatmullRom, FOVTransition, formatNumber, formatProgressText, getAnchorOffset, getAudioContext, getDamageNumberColor, getNotificationColor, getNotificationIcon, getTextDirection, HapticFeedback, Howl, Howler, InputManager, InputStateMachine, isAudioContextUnlocked, lerp, lerpVector3, normalizeAxisValue, resumeAudioContext, SoundManager, SpatialAudio, screenToWorld, setupAutoUnlock, slerp, smoothDamp, smoothDampVector3, suspendAudioContext, uiEaseOutCubic, uiEaseOutElastic, uiLerp, unlockAudioContext, worldToScreen, } from '../core';
