// @jbcom/strata/dist/api/rendering.js
/**
 * @module Rendering
 * @category Rendering Pipeline
 *
 * Rendering Pipeline - Shaders, Post-Processing, and Materials
 *
 * Low-level rendering tools for graphics programmers - custom shaders,
 * post-processing effects, and advanced materials.
 *
 * @example
 * ```tsx
 * import { CinematicEffects, ToonMesh, DissolveMesh } from '@jbcom/strata/api/rendering';
 *
 * function StylizedScene() {
 *   return (
 *     <>
 *       <CinematicEffects intensity={0.8} />
 *       <ToonMesh geometry={geometry} outlineWidth={0.02} />
 *       <DissolveMesh progress={0.5} edgeColor="cyan" />
 *     </>
 *   );
 * }
 * ```
 */
// Post-Processing Effects - React components
// Stylized Shaders
// Ray Marching - React component
export { CinematicEffects, CrystalMesh, DissolveMesh, DreamyEffects, DynamicDOF, EffectStack, Forcefield, GlitchMesh, GradientMesh, HologramMesh, HorrorEffects, MotionBlurEffect, NeonEffects, Outline, Raymarching, RealisticEffects, ToonMesh, VintageEffects, } from '../components';
// Post-Processing Effects - Core utilities
// Ray Marching - Core utilities
export { apertureToBokehScale, blendPostProcessingPresets, calculateFocusDistance, calculateFocusDistanceToMesh, createRaymarchingGeometry, createRaymarchingMaterial, defaultEffectSettings, dofScenarios, focalLengthToFOV, fovToFocalLength, getTimeOfDayEffects, lutConfigs, } from '../core';
// GLSL Shaders (raw)
export * from '../shaders';
// Utilities
export * from '../utils';
