// @jbcom/strata/dist/index.js
/**
 * @jbcom/strata
 *
 * The complete solution for foreground, midground, and background layer
 * 3D gaming in Node.js. Provides terrain, water, vegetation, characters,
 * fur, shells, molecular rendering, and more.
 *
 * Organized into presets for easy integration into your game.
 */
// React components
export { AdvancedWater, AmbientAudio, AnimatedBillboard, AudioEmitter, AudioEnvironment, AudioListener, 
// Audio components
AudioProvider, AudioZone, Billboard, BlinkController, BreathingAnimation, Buoyancy, CameraShake, 
// Physics components
CharacterController, CinematicCamera, CinematicEffects, CloudLayer, CloudSky, Crosshair, CrystalMesh, createTimeOfDay, DamageNumber, 
// Decal components
Decal, DecalPool, Destructible, DialogBox, DissolveMesh, DreamyEffects, DynamicDOF, 
// Post-Processing components
EffectStack, EnhancedFog, 
// Camera components
FollowCamera, FootstepAudio, Forcefield, FPSCamera, GameStateContext, 
// State management components
GameStateProvider, GlitchMesh, 
// God rays components
GodRays, GPUInstancedMesh, GradientMesh, GrassInstances, GroundSwitch, generateInstanceData, HeadTracker, 
// UI components
HealthBar, HologramMesh, HorrorEffects, 
// Animation components
IKChain, IKLimb, Impostor, Inventory, 
// Input components
Joystick3D, Lightning, LightShafts, LODGroup, 
// LOD components
LODMesh, LODVegetation, LookAt, Minimap, MotionBlurEffect, Nameplate, NeonEffects, Notification, OrbitCamera, Outline, ParticleBurst, ParticleEmitter, PersistGate, PositionalAudio, PressurePlate, ProceduralSky, ProceduralWalk, ProgressBar3D, Ragdoll, Rain, Raymarching, RealisticEffects, RockInstances, Snow, SpringBone, StateDebugger, syncYukaToThree, TailPhysics, Tooltip, 
// Shader components
ToonMesh, TreeInstances, TriggerComposer, threeVector3ToYuka, UnderwaterOverlay, useAudioContext, useAudioManager, useAutoSave, useCameraTransition, useCheckpoint, useGameState, useGameStateContext, useSaveLoad, useUndo, useYukaContext, VehicleBody, VintageEffects, VolumetricClouds, VolumetricEffects, VolumetricFogMesh, VolumetricPointLight, VolumetricSpotlight, WallButton, Water, WeatherAudio, WeatherEffects, 
// AI components (YukaJS integration)
YukaEntityManager, YukaNavMesh, YukaPath, YukaStateMachine, YukaVehicle, yukaVector3ToThree, } from './components';
// Core algorithms (pure TypeScript, no React)
export { adaptCloudColorsForTimeOfDay, angleToAxis, 
// Re-export with original name for convenience; aliased in core to avoid potential collisions
animationClampAngle as clampAngle, apertureToBokehScale, applyDrag, applySpriteSheetFrame, axisToAngle, axisToMagnitude, batchLODObjects, blendGodRayColors, blendPostProcessingPresets, CameraShake as CoreCameraShake, CCDSolver, 
// Physics (pure TS)
CollisionLayer, calcNormal, calculateBoneRotation, calculateBuoyancyForce, calculateChecksum, calculateExplosionForce, calculateFade, 
// Post-Processing utilities (pure TS)
calculateFocusDistance, calculateFocusDistanceToMesh, calculateForce, calculateGodRayIntensityFromAngle, calculateHeadBob, calculateImpostorAngle, calculateImpulse, calculateJumpImpulse, calculateLandingVelocity, calculateLODLevel, calculateLookAhead, calculateRadialBlur, calculateScatteringIntensity, calculateScreenShakeIntensity, calculateScreenSpaceSize, calculateSlopeAngle, calculateSteeringAngle, calculateSunOcclusion, calculateSuspensionForce, calculateTemperature, calculateVegetationDensity, calculateWindOffset, clampAxis, clampProgress, collisionFilters, createAdvancedWaterMaterial, createBillboardMatrix, createBloodSplatterTexture, createBoneChain, createBoneChainFromLengths, createBulletHoleTexture, createCloudLayerGeometry, 
// Clouds (pure TS)
createCloudLayerMaterial, createDecalTexture, createDefaultBuoyancyConfig, createDefaultCharacterConfig, createDefaultCloudSkyConfig, createDefaultCrosshair, createDefaultDamageNumber, createDefaultDestructibleConfig, createDefaultDialog, createDefaultInventory, createDefaultMinimap, createDefaultNameplate, createDefaultNotification, createDefaultPhysicsConfig, createDefaultProgressBar, createDefaultTooltip, createDefaultVehicleConfig, createDitherPattern, createFootprintTexture, 
// State management (pure TS)
// Note: Zustand's create, useStore, temporal, and immer are not re-exported.
// Import directly from 'zustand' and related packages if needed.
createGameStore, createGeometryFromMarchingCubes, 
// God rays (pure TS)
createGodRaysMaterial, createHumanoidRagdoll, createImpostorGeometry, createImpostorTexture, createInputManager, createInstancedMesh, createLODLevels, createParticleEmitter, createPersistenceAdapter, createPointLightSphereGeometry, createRaymarchingGeometry, 
// Ray marching (pure TS)
createRaymarchingMaterial, createScorchMarkTexture, createSkyGeometry, 
// Sky (pure TS)
createSkyMaterial, createSoundManager, createSpatialAudio, createSpotlightConeGeometry, createSpriteSheetAnimation, createSpriteSheetMaterial, createUnderwaterOverlayMaterial, createVegetationLODLevels, createVolumetricCloudGeometry, createVolumetricCloudMaterial, 
// Volumetrics (pure TS)
createVolumetricFogMeshMaterial, createVolumetricPointLightMaterial, createVolumetricSpotlightMaterial, createWaterGeometry, 
// Water (pure TS)
createWaterMaterial, createWaterPuddleTexture, createWeatherSystem, createWebPersistenceAdapter, createWindSimulation, DEFAULT_SPATIAL_CONFIG, 
// Decals and billboards (pure TS)
DecalProjector, dampedSpring, dampedSpringVector3, defaultEffectSettings, dofScenarios, ENVIRONMENT_PRESETS, easeInCubic, easeInOutCubic, easeOutCubic, easeOutElastic, evaluateCatmullRom, 
// Animation (pure TS)
FABRIKSolver, FOVTransition, fbm, fbmNoise2D, focalLengthToFOV, formatNumber, formatProgressText, fovToFocalLength, generateDebrisVelocity, 
// Instancing (pure TS)
generateInstanceData as generateInstanceDataCore, generateLODGeometries, generateTerrainChunk, 
// UI utilities (pure TS)
// Note: lerp, easeOutCubic, easeOutElastic are available above from camera utilities.
getAnchorOffset, getAudioContext, getBiomeAt, getDamageNumberColor, getLightScreenPosition, getNotificationColor, getNotificationIcon, getPrecipitationType, getSpriteSheetUVs, getTerrainHeight, getTextDirection, getTimeOfDayEffects, HapticFeedback, hermiteInterpolate, 
// Input (pure TS)
InputManager, InputStateMachine, interpolateLODMaterials, isAudioContextUnlocked, isWalkableSlope, 
// LOD (pure TS)
LODManager, LookAtController, 
// Camera utilities (pure TS)
lerp, lerpVector3, lutConfigs, 
// Marching cubes
marchingCubes, noise3D, normalizeAxisValue, opIntersection, opSmoothIntersection, opSmoothSubtraction, opSmoothUnion, opSubtraction, opUnion, 
// Particles (pure TS)
ParticleEmitter as CoreParticleEmitter, ProceduralGait, projectVelocityOntoGround, resumeAudioContext, 
// Audio (pure TS)
// Note: Howl and Howler from howler.js are not re-exported.
// Import directly from 'howler' if needed.
SoundManager, SpatialAudio, SpringChain, SpringDynamics, sampleCloudDensity, sampleCurve, screenToWorld, sdBox, sdCapsule, sdCaves, sdCone, sdPlane, sdRock, 
// SDF
sdSphere, sdTerrain, sdTorus, setupAutoUnlock, shouldUseLOD, simplifyGeometry, slerp, smoothDamp, smoothDampVector3, sortBillboardsByDepth, suspendAudioContext, TwoBoneIKSolver, unlockAudioContext, updateBillboardRotation, updateGodRaysLightPosition, updateImpostorUV, updateSpriteSheetAnimation, verifyChecksum, 
// Weather (pure TS)
WeatherSystem as CoreWeatherSystem, WindSimulation, warpedFbm, webPersistenceAdapter, worldToScreen, } from './core';
// Hooks (Yuka steering behaviors)
export * from './hooks';
// Hooks
export { useAlignment, useArrive, useCohesion, useEvade, useFlee, useFollowPath, useInterpose, useObstacleAvoidance, useOffsetPursuit, usePursue, useSeek, useSeparation, useWander, } from './hooks';
// Presets (organized by layer: background, midground, foreground)
export * from './presets';
// GLSL shaders
export * from './shaders';
// Utilities
export * from './utils';
