// @jbcom/strata/dist/core/index.js
/**
 * Core algorithm exports
 */
// Animation system
export { CCDSolver, calculateBoneRotation, clampAngle, 
// Alias to avoid potential collision with camera utilities
clampAngle as animationClampAngle, createBoneChain, createBoneChainFromLengths, dampedSpring, dampedSpringVector3, FABRIKSolver, hermiteInterpolate, LookAtController, ProceduralGait, SpringChain, SpringDynamics, sampleCurve, TwoBoneIKSolver, } from './animation';
// Audio system
export { createSoundManager, createSpatialAudio, DEFAULT_SPATIAL_CONFIG, ENVIRONMENT_PRESETS, getAudioContext, Howl, Howler, isAudioContextUnlocked, resumeAudioContext, SoundManager, SpatialAudio, setupAutoUnlock, suspendAudioContext, unlockAudioContext, } from './audio';
// Camera utilities
export { CameraShake, calculateHeadBob, calculateLookAhead, calculateScreenShakeIntensity, easeInCubic, easeInOutCubic, easeOutCubic, easeOutElastic, evaluateCatmullRom, FOVTransition, lerp, lerpVector3, slerp, smoothDamp, smoothDampVector3, } from './camera';
export * from './clouds';
// Decals
export * from './decals';
// Decals and billboards
export { applySpriteSheetFrame, createBillboardMatrix, createBloodSplatterTexture, createBulletHoleTexture, createDecalTexture, createFootprintTexture, createScorchMarkTexture, createSpriteSheetAnimation, createSpriteSheetMaterial, createWaterPuddleTexture, DecalProjector, getSpriteSheetUVs, sortBillboardsByDepth, updateBillboardRotation, updateSpriteSheetAnimation, } from './decals';
// ECS module
export * from './ecs';
export * from './godRays';
// Input system
export { angleToAxis, axisToAngle, axisToMagnitude, clampAxis, createInputManager, HapticFeedback, InputManager, InputStateMachine, normalizeAxisValue, } from './input';
// Core modules (pure TypeScript, no React)
export { createInstancedMesh, generateInstanceData } from './instancing';
// LOD
export * from './lod';
// LOD (Level of Detail)
export { batchLODObjects, calculateImpostorAngle, calculateLODLevel, calculateScreenSpaceSize, calculateVegetationDensity, createDitherPattern, createImpostorGeometry, createImpostorTexture, createLODLevels, createVegetationLODLevels, generateLODGeometries, interpolateLODMaterials, LODManager, shouldUseLOD, simplifyGeometry, updateImpostorUV, } from './lod';
// Marching cubes
export { createGeometryFromMarchingCubes, generateTerrainChunk, marchingCubes, } from './marching-cubes';
// Particles (pure TypeScript)
export { createParticleEmitter, ParticleEmitter } from './particles';
// Physics system
export { applyDrag, CollisionLayer, calculateBuoyancyForce, calculateExplosionForce, calculateForce, calculateImpulse, calculateJumpImpulse, calculateLandingVelocity, calculateSlopeAngle, calculateSteeringAngle, calculateSuspensionForce, collisionFilters, createDefaultBuoyancyConfig, createDefaultCharacterConfig, createDefaultDestructibleConfig, createDefaultPhysicsConfig, createDefaultVehicleConfig, createHumanoidRagdoll, generateDebrisVelocity, isWalkableSlope, projectVelocityOntoGround, } from './physics';
// Post-Processing utilities
export { apertureToBokehScale, blendPostProcessingPresets, calculateFocusDistance, calculateFocusDistanceToMesh, defaultEffectSettings, dofScenarios, focalLengthToFOV, fovToFocalLength, getTimeOfDayEffects, lutConfigs, } from './postProcessing';
export * from './raymarching';
// SDF primitives and operations
export { 
// Utilities
calcNormal, fbm, 
// Terrain
getBiomeAt, getTerrainHeight, 
// Noise functions
noise3D, opIntersection, opSmoothIntersection, opSmoothSubtraction, opSmoothUnion, opSubtraction, 
// Boolean operations
opUnion, sdBox, sdCapsule, sdCaves, sdCone, sdPlane, sdRock, 
// Primitives
sdSphere, sdTerrain, sdTorus, warpedFbm, } from './sdf';
// Shader utilities
export { animationSnippet, buildFragmentShader, buildVertexShader, colorSnippet, composeShaderChunks, createColorUniform, createProgressUniform, createTimeUniform, createVector2Uniform, createVector3Uniform, lightingSnippet, noiseSnippet, ShaderChunks, } from './shaders';
// Shared utilities
export * from './shared';
export * from './sky';
// State management system (Zustand-based)
export { calculateChecksum, create, createGameStore, createPersistenceAdapter, createWebPersistenceAdapter, immer, temporal, useStore, verifyChecksum, webPersistenceAdapter, } from './state';
// UI system
export { calculateFade, clampProgress, createDefaultCrosshair, createDefaultDamageNumber, createDefaultDialog, createDefaultInventory, createDefaultMinimap, createDefaultNameplate, createDefaultNotification, createDefaultProgressBar, createDefaultTooltip, easeOutCubic as uiEaseOutCubic, easeOutElastic as uiEaseOutElastic, formatNumber, formatProgressText, getAnchorOffset, getDamageNumberColor, getNotificationColor, getNotificationIcon, getTextDirection, 
// UI-specific aliases for lerp/easing functions
lerp as uiLerp, screenToWorld, worldToScreen, } from './ui';
export * from './volumetrics';
export * from './water';
// Weather system
export { calculateTemperature, createWeatherSystem, createWindSimulation, getPrecipitationType, WeatherSystem, WindSimulation, } from './weather';
