// @jbcom/strata/dist/core/sky.js
/**
 * Sky Materials - Core TypeScript (no React)
 *
 * Pure TypeScript functions for creating sky materials
 */
import * as THREE from 'three';
import { createSkyUniforms, skyFragmentShader, skyVertexShader } from '../shaders/sky';
/**
 * Create sky material (pure TypeScript)
 */
export function createSkyMaterial(options) {
    const { timeOfDay, weather = {}, gyroTilt, time = 0 } = options;
    // Input validation
    if (timeOfDay &&
        timeOfDay.sunIntensity !== undefined &&
        (timeOfDay.sunIntensity < 0 || timeOfDay.sunIntensity > 1)) {
        throw new Error('createSkyMaterial: sunIntensity must be between 0 and 1');
    }
    if (timeOfDay &&
        timeOfDay.sunAngle !== undefined &&
        (timeOfDay.sunAngle < 0 || timeOfDay.sunAngle > 180)) {
        throw new Error('createSkyMaterial: sunAngle must be between 0 and 180');
    }
    const defaultTimeOfDay = {
        sunIntensity: 1.0,
        sunAngle: 60,
        ambientLight: 0.8,
        starVisibility: 0,
        fogDensity: 0,
    };
    const defaultWeather = {
        intensity: 0,
    };
    const mergedTimeOfDay = { ...defaultTimeOfDay, ...timeOfDay };
    const mergedWeather = { ...defaultWeather, ...weather };
    const uniforms = createSkyUniforms(mergedTimeOfDay, mergedWeather, gyroTilt);
    uniforms.uTime.value = time;
    return new THREE.ShaderMaterial({
        uniforms,
        vertexShader: skyVertexShader,
        fragmentShader: skyFragmentShader,
        side: THREE.DoubleSide,
    });
}
/**
 * Create sky geometry (pure TypeScript)
 */
export function createSkyGeometry(size = [200, 100]) {
    return new THREE.PlaneGeometry(size[0], size[1], 1, 1);
}
