// @jbcom/strata/dist/core/audio/types.js
/**
 * Audio System Type Definitions
 *
 * Types for 2D audio (Howler) and 3D spatial audio (Three.js).
 *
 * @module core/audio
 * @public
 */
/**
 * Predefined environment effect configurations.
 */
export const ENVIRONMENT_PRESETS = {
    outdoor: { preset: 'outdoor', wetLevel: 0.1, dryLevel: 0.9, decay: 0.5 },
    indoor: { preset: 'indoor', wetLevel: 0.3, dryLevel: 0.7, decay: 1.5 },
    cave: { preset: 'cave', wetLevel: 0.5, dryLevel: 0.5, decay: 4.0 },
    underwater: { preset: 'underwater', wetLevel: 0.4, dryLevel: 0.6, decay: 2.0 },
    hall: { preset: 'hall', wetLevel: 0.4, dryLevel: 0.6, decay: 2.5 },
    studio: { preset: 'studio', wetLevel: 0.2, dryLevel: 0.8, decay: 0.8 },
    none: { preset: 'none', wetLevel: 0, dryLevel: 1, decay: 0 },
};
/**
 * Default spatial audio configuration.
 */
export const DEFAULT_SPATIAL_CONFIG = {
    refDistance: 1,
    maxDistance: 10000,
    rolloffFactor: 1,
    distanceModel: 'inverse',
    coneInnerAngle: 360,
    coneOuterAngle: 360,
    coneOuterGain: 0,
    panningModel: 'HRTF',
};
