// @jbcom/strata/dist/presets/index.js
/**
 * Strata Presets - Organized game development primitives
 *
 * This module exports all preset systems organized by layer:
 * - Background: Sky, volumetrics, terrain
 * - Midground: Water, vegetation, instancing
 * - Foreground: Characters, fur, shells, molecular
 */
export { createInstancedMesh, generateInstanceData } from '../core/instancing';
export * from '../core/marching-cubes';
export * from '../core/raymarching';
export { calcNormal, fbm, getBiomeAt, getTerrainHeight, noise3D, opIntersection, opSmoothIntersection, opSmoothSubtraction, opSmoothUnion, opSubtraction, opUnion, sdBox, sdCapsule, sdCaves, sdCone, sdPlane, sdRock, sdSphere, sdTerrain, sdTorus, warpedFbm, } from '../core/sdf';
// Background Layer
export * from '../core/sky';
export * from '../core/volumetrics';
// AI presets
export { createFlock, createFlockMemberPreset, createFollowerPreset, createGuardPreset, createPredatorPreset, createPreyPreset, } from './ai';
// Animation presets
export { BOUNCY_SPRING_PRESET, blendGaitPresets, blendSpringPresets, CLOTH_SPRING_PRESET, CRAWL_GAIT_PRESET, createCustomGaitPreset, createCustomIKPreset, createCustomSpringPreset, FINGER_IK_PRESET, FLOPPY_SPRING_PRESET, gaitPresets, getGaitPreset, getIKPreset, getLookAtPreset, getSpringPreset, HAIR_SPRING_PRESET, HUMAN_ARM_IK_PRESET, HUMAN_LEG_IK_PRESET, ikPresets, JELLY_SPRING_PRESET, LAZY_LOOKAT_PRESET, LIMP_GAIT_PRESET, lookAtPresets, MARCH_GAIT_PRESET, ORGANIC_LOOKAT_PRESET, ROBOTIC_LOOKAT_PRESET, RUN_GAIT_PRESET, SMOOTH_LOOKAT_PRESET, SNAPPY_LOOKAT_PRESET, SNEAK_GAIT_PRESET, SPIDER_LEG_IK_PRESET, SPINE_IK_PRESET, STIFF_SPRING_PRESET, springPresets, TAIL_IK_PRESET, TENTACLE_IK_PRESET, WALK_GAIT_PRESET, } from './animation';
// Audio presets
export { audioPresets, calculateWeatherAudioIntensity, combatSoundPresets, createCustomAudioPreset, footstepPresets, getAudioPreset, getCombatSoundPreset, getFootstepPreset, spatialAudioConfigs, weatherAudioPresets, } from './audio';
export * from './billboards';
// Camera presets
export * from './camera';
export * from './characters';
// Clouds
export * from './clouds';
export * from './decals';
// Foreground Layer
export * from './fur';
// Input presets
export * from './input';
export * from './lighting';
// LOD presets
export * from './lod';
export * from './molecular';
export * from './particles';
// Physics presets
export { BOAT_BUOYANCY_PRESET, BOUNCY_MATERIAL_PRESET, buoyancyPresets, CAR_VEHICLE_PRESET, CONCRETE_MATERIAL_PRESET, characterPresets, destructiblePresets, FPS_CHARACTER_PRESET, GLASS_PRESET, getBuoyancyPreset, getCharacterPreset, getDestructiblePreset, getMaterialPreset, getVehiclePreset, HEAVY_BUOYANCY_PRESET, ICE_MATERIAL_PRESET, LIGHT_BUOYANCY_PRESET, MEDIUM_BUOYANCY_PRESET, METAL_MATERIAL_PRESET, MOTORCYCLE_VEHICLE_PRESET, MUD_MATERIAL_PRESET, materialPresets, PLATFORMER_CHARACTER_PRESET, RUBBER_MATERIAL_PRESET, SPORTS_CAR_PRESET, STONE_PRESET, TANK_CHARACTER_PRESET, THIRD_PERSON_CHARACTER_PRESET, TRUCK_VEHICLE_PRESET, vehiclePresets, WOOD_MATERIAL_PRESET, WOODEN_CRATE_PRESET, } from './physics';
export * from './postprocessing';
export * from './reflections';
// Shader presets
export { colorPalettes, crystalPresets, dissolvePresets, forcefieldPresets, getColorPalette, getCrystalPreset, getDissolvePreset, getForcefieldPreset, getGlitchPreset, getGradientPreset, getHologramPreset, getScanlinePreset, getToonPreset, glitchPresets, gradientPresets, hologramPresets, scanlinePresets, toonPresets, } from './shaders';
// Lighting & Effects
export * from './shadows';
// State presets
export { ALL_STATE_PRESETS, AUTOSAVE_CONFIG_DISABLED, AUTOSAVE_CONFIG_FREQUENT, AUTOSAVE_CONFIG_INFREQUENT, AUTOSAVE_CONFIG_MODERATE, addExperience, addInventoryItem, COUNTER_STATE_PRESET, collectCoin, completeQuest, createPlatformerState, createPuzzleState, createRPGState, createSandboxState, DEFAULT_COUNTER_STATE, DEFAULT_PLATFORMER_STATE, DEFAULT_PUZZLE_STATE, DEFAULT_RPG_STATE, DEFAULT_SANDBOX_STATE, getStatePreset, loseLife, PLATFORMER_STATE_PRESET, PUZZLE_STATE_PRESET, placeBlock, RPG_STATE_PRESET, removeBlock, SANDBOX_STATE_PRESET, unlockLevel, } from './state';
export * from './terrain';
// UI presets
export { CIRCLE_CROSSHAIR, CIRCULAR_MINIMAP, COMIC_DIALOG, COMPASS_MINIMAP, createCustomDialog, createCustomHealthBar, createCustomInventory, crosshairPresets, DEFAULT_CROSSHAIR, DOT_CROSSHAIR, dialogPresets, ERROR_NOTIFICATION, FPS_HEALTH_BAR, GRID_INVENTORY, getCrosshairPreset, getDialogPreset, getHealthBarPreset, getInventoryPreset, getMinimapPreset, getNotificationPreset, HOTBAR_INVENTORY, healthBarPresets, INFO_NOTIFICATION, inventoryPresets, LIST_INVENTORY, MINIMALIST_HEALTH_BAR, MMO_HEALTH_BAR, minimapPresets, notificationPresets, QUEST_NOTIFICATION, RADAR_MINIMAP, RETRO_HEALTH_BAR, RPG_DIALOG, RPG_HEALTH_BAR, SNIPER_CROSSHAIR, SQUARE_MINIMAP, SUBTITLE_DIALOG, SUCCESS_NOTIFICATION, TACTICAL_CROSSHAIR, VISUAL_NOVEL_DIALOG, WARNING_NOTIFICATION, WHEEL_INVENTORY, } from './ui';
export * from './vegetation';
// Midground Layer
export * from './water';
// Weather
export * from './weather';
