// @jbcom/strata/dist/core/state/adapters/web/persistence.js
/**
 * Web LocalStorage Persistence Adapter
 *
 * Implements PersistenceAdapter interface for web browsers using localStorage.
 *
 * @module core/state/adapters/web/persistence
 */
export class WebPersistenceAdapter {
    async save(key, data) {
        try {
            if (typeof localStorage === 'undefined') {
                console.warn('localStorage is not available');
                return false;
            }
            const jsonString = JSON.stringify(data);
            localStorage.setItem(key, jsonString);
            return true;
        }
        catch (error) {
            console.error('Failed to save state:', error);
            return false;
        }
    }
    async load(key) {
        try {
            if (typeof localStorage === 'undefined') {
                return null;
            }
            const jsonString = localStorage.getItem(key);
            if (!jsonString)
                return null;
            return JSON.parse(jsonString);
        }
        catch (error) {
            console.error('Failed to load state:', error);
            return null;
        }
    }
    async delete(key) {
        try {
            if (typeof localStorage === 'undefined') {
                return false;
            }
            localStorage.removeItem(key);
            return true;
        }
        catch (error) {
            console.error('Failed to delete save:', error);
            return false;
        }
    }
    async listSaves(prefix) {
        const saves = [];
        if (typeof localStorage === 'undefined') {
            return saves;
        }
        try {
            for (let i = 0; i < localStorage.length; i++) {
                const key = localStorage.key(i);
                // Check for prefix with underscore to avoid matching unrelated keys
                const prefixWithSeparator = `${prefix}_`;
                if (key?.startsWith(prefixWithSeparator)) {
                    saves.push(key.slice(prefixWithSeparator.length));
                }
            }
        }
        catch (error) {
            console.error('Failed to list saves:', error);
        }
        return saves;
    }
    async getSaveInfo(key) {
        try {
            const saveData = await this.load(key);
            if (!saveData)
                return null;
            return {
                timestamp: saveData.timestamp,
                version: saveData.version,
            };
        }
        catch (error) {
            console.error('Failed to get save info:', error);
            return null;
        }
    }
}
export const webPersistenceAdapter = new WebPersistenceAdapter();
export function createWebPersistenceAdapter() {
    return new WebPersistenceAdapter();
}
