// @miniplex/bucket/dist/miniplex-bucket.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var eventery = require('eventery');

function _arrayLikeToArray(arr, len) {
  if (len == null || len > arr.length) len = arr.length;
  for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i];
  return arr2;
}

function _unsupportedIterableToArray(o, minLen) {
  if (!o) return;
  if (typeof o === "string") return _arrayLikeToArray(o, minLen);
  var n = Object.prototype.toString.call(o).slice(8, -1);
  if (n === "Object" && o.constructor) n = o.constructor.name;
  if (n === "Map" || n === "Set") return Array.from(o);
  if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen);
}

function _createForOfIteratorHelper(o, allowArrayLike) {
  var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"];
  if (!it) {
    if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") {
      if (it) o = it;
      var i = 0;
      var F = function () {};
      return {
        s: F,
        n: function () {
          if (i >= o.length) return {
            done: true
          };
          return {
            done: false,
            value: o[i++]
          };
        },
        e: function (e) {
          throw e;
        },
        f: F
      };
    }
    throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
  }
  var normalCompletion = true,
    didErr = false,
    err;
  return {
    s: function () {
      it = it.call(o);
    },
    n: function () {
      var step = it.next();
      normalCompletion = step.done;
      return step;
    },
    e: function (e) {
      didErr = true;
      err = e;
    },
    f: function () {
      try {
        if (!normalCompletion && it.return != null) it.return();
      } finally {
        if (didErr) throw err;
      }
    }
  };
}

function _classCallCheck(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
}

function _toPrimitive(input, hint) {
  if (typeof input !== "object" || input === null) return input;
  var prim = input[Symbol.toPrimitive];
  if (prim !== undefined) {
    var res = prim.call(input, hint || "default");
    if (typeof res !== "object") return res;
    throw new TypeError("@@toPrimitive must return a primitive value.");
  }
  return (hint === "string" ? String : Number)(input);
}

function _toPropertyKey(arg) {
  var key = _toPrimitive(arg, "string");
  return typeof key === "symbol" ? key : String(key);
}

function _defineProperties(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || false;
    descriptor.configurable = true;
    if ("value" in descriptor) descriptor.writable = true;
    Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor);
  }
}
function _createClass(Constructor, protoProps, staticProps) {
  if (protoProps) _defineProperties(Constructor.prototype, protoProps);
  if (staticProps) _defineProperties(Constructor, staticProps);
  Object.defineProperty(Constructor, "prototype", {
    writable: false
  });
  return Constructor;
}

function _defineProperty(obj, key, value) {
  key = _toPropertyKey(key);
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }
  return obj;
}

var _Symbol$iterator;

/**
 * A class wrapping an array of entities of a specific type, providing
 * performance-optimized methods for adding, looking up and removing entities, and events
 * for when entities are added or removed.
 */
_Symbol$iterator = Symbol.iterator;
var Bucket = /*#__PURE__*/function () {
  function Bucket() {
    var _entities = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
    _classCallCheck(this, Bucket);
    /* VERSIONING */
    _defineProperty(this, "_version", 0);
    /**
     * Fired when an entity has been added to the bucket.
     */
    _defineProperty(this, "onEntityAdded", new eventery.Event());
    /**
     * Fired when an entity is about to be removed from the bucket.
     */
    _defineProperty(this, "onEntityRemoved", new eventery.Event());
    /**
     * A map of entity positions, used for fast lookups.
     */
    _defineProperty(this, "entityPositions", new Map());
    this._entities = _entities;
    this.add = this.add.bind(this);
    this.remove = this.remove.bind(this);

    /* Register all entity positions */
    for (var i = 0; i < _entities.length; i++) {
      this.entityPositions.set(_entities[i], i);
    }
  }
  _createClass(Bucket, [{
    key: "version",
    get:
    /**
     * The current version of the bucket. Increases every time an entity is
     * added or removed.
     */
    function get() {
      return this._version;
    }

    /**
     * An array of all entities within the bucket. Please note that for iterating
     * over the entities in this bucket, it is recommended that you use the
     * `for (const entity of bucket)` iterator form.
     */
  }, {
    key: "entities",
    get: function get() {
      return this._entities;
    }

    /* Custom iterator that iterates over all entities in reverse order. */
  }, {
    key: _Symbol$iterator,
    value: function value() {
      var _this = this;
      var index = this._entities.length;
      var result = {
        value: undefined,
        done: false
      };
      return {
        next: function next() {
          result.value = _this._entities[--index];
          result.done = index < 0;
          return result;
        }
      };
    }
  }, {
    key: "size",
    get:
    /**
     * Returns the total size of the bucket, i.e. the number of entities it contains.
     */
    function get() {
      return this.entities.length;
    }

    /**
     * Returns the first entity in the bucket, or `undefined` if the bucket is empty.
     */
  }, {
    key: "first",
    get: function get() {
      return this.entities[0];
    }

    /**
     * Returns true if the bucket contains the given entity.
     *
     * @param entity The entity to check for.
     * @returns `true` if the specificed entity is in this bucket, `false` otherwise.
     */
  }, {
    key: "has",
    value: function has(entity) {
      return this.entityPositions.has(entity);
    }

    /**
     * Adds the given entity to the bucket. If the entity is already in the bucket, it is
     * not added again.
     *
     * @param entity The entity to add to the bucket.
     * @returns The entity passed into this function (regardless of whether it was added or not).
     */
  }, {
    key: "add",
    value: function add(entity) {
      if (entity && !this.has(entity)) {
        this.entities.push(entity);
        this.entityPositions.set(entity, this.entities.length - 1);

        /* Increase version */
        this._version++;

        /* Emit our own onEntityAdded event */
        this.onEntityAdded.emit(entity);
      }
      return entity;
    }

    /**
     * Removes the given entity from the bucket. If the entity is not in the bucket, nothing
     * happens.
     *
     * @param entity The entity to remove from the bucket.
     * @returns The entity passed into this function (regardless of whether it was removed or not).
     */
  }, {
    key: "remove",
    value: function remove(entity) {
      /* TODO: Return early if entity is not in bucket. */
      if (this.has(entity)) {
        /* Emit our own onEntityRemoved event. */
        this.onEntityRemoved.emit(entity);

        /* Get the entity's current position. */
        var index = this.entityPositions.get(entity);
        this.entityPositions["delete"](entity);

        /* Perform shuffle-pop if there is more than one entity. */
        var other = this.entities[this.entities.length - 1];
        if (other !== entity) {
          this.entities[index] = other;
          this.entityPositions.set(other, index);
        }

        /* Remove the entity from the entities array. */
        this.entities.pop();

        /* Bump version */
        this._version++;
      }
      return entity;
    }

    /**
     * Removes all entities from the bucket. Will cause the `onEntityRemoved` event to be
     * fired for each entity.
     */
  }, {
    key: "clear",
    value: function clear() {
      var _iterator = _createForOfIteratorHelper(this),
        _step;
      try {
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          var _entity = _step.value;
          this.remove(_entity);
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }
    }
  }]);
  return Bucket;
}();

exports.Bucket = Bucket;
