// @jbcom/strata/dist/components/Sky.js
import { jsx as _jsx } from "react/jsx-runtime";
/**
 * Procedural Sky component
 *
 * Lifted from Otterfall biome selector diorama.
 */
import { useFrame } from '@react-three/fiber';
import { useEffect, useMemo, useRef } from 'react';
import { createSkyGeometry, createSkyMaterial } from '../core/sky';
const defaultTimeOfDay = {
    sunIntensity: 1.0,
    sunAngle: 60,
    ambientLight: 0.8,
    starVisibility: 0,
    fogDensity: 0,
};
const defaultWeather = {
    intensity: 0,
};
/**
 * Procedural sky with day/night cycle, stars, clouds, and weather effects
 */
export function ProceduralSky({ timeOfDay: timeOfDayProp = {}, weather: weatherProp = {}, size = [200, 100], distance = 50, }) {
    const meshRef = useRef(null);
    const timeOfDay = { ...defaultTimeOfDay, ...timeOfDayProp };
    const weather = { ...defaultWeather, ...weatherProp };
    const material = useMemo(() => {
        return createSkyMaterial({
            timeOfDay,
            weather,
        });
    }, [timeOfDay, weather]);
    const geometry = useMemo(() => {
        return createSkyGeometry(size);
    }, [size]);
    useFrame((state) => {
        if (meshRef.current && material.uniforms) {
            material.uniforms.uTime.value = state.clock.elapsedTime;
            material.uniforms.uSunIntensity.value = timeOfDay.sunIntensity;
            material.uniforms.uSunAngle.value = timeOfDay.sunAngle;
            material.uniforms.uAmbientLight.value = timeOfDay.ambientLight;
            material.uniforms.uStarVisibility.value = timeOfDay.starVisibility;
            material.uniforms.uFogDensity.value = timeOfDay.fogDensity;
            material.uniforms.uWeatherIntensity.value = weather.intensity;
            // Subtle gyroscopic effect
            const tiltX = Math.sin(state.clock.elapsedTime * 0.1) * 0.02;
            const tiltY = Math.cos(state.clock.elapsedTime * 0.15) * 0.02;
            material.uniforms.uGyroTilt.value.set(tiltX, tiltY);
        }
    });
    useEffect(() => {
        return () => {
            material.dispose();
            geometry.dispose();
        };
    }, [material, geometry]);
    return (_jsx("mesh", { ref: meshRef, position: [0, 0, -distance], geometry: geometry, children: _jsx("primitive", { object: material, attach: "material" }) }));
}
/**
 * Convenience function to create time of day state from hour
 */
export function createTimeOfDay(hour) {
    // Normalize to 0-24
    const normalizedHour = ((hour % 24) + 24) % 24;
    // Sun angle: peaks at noon (90°), 0 at 6am/6pm
    const sunAngle = Math.max(0, Math.sin(((normalizedHour - 6) / 12) * Math.PI) * 90);
    // Sun intensity based on time
    let sunIntensity = 0;
    if (normalizedHour >= 6 && normalizedHour <= 18) {
        sunIntensity = Math.sin(((normalizedHour - 6) / 12) * Math.PI);
    }
    // Star visibility (inverse of sun)
    const starVisibility = Math.max(0, 1 - sunIntensity * 2);
    // Ambient light
    const ambientLight = 0.2 + sunIntensity * 0.6;
    return {
        sunIntensity,
        sunAngle,
        ambientLight,
        starVisibility,
        fogDensity: 0,
    };
}
